### ISSUE TITLE
Zstandard

### ISSUE THREAD

**Issue by 89z** [rust-lang/rust#82595]
I noticed that the files here:

https://static.rust-lang.org/dist/channel-rust-stable.toml

are available with Tar.Gz and Tar.Xz. I was wondering if Zstandard has been considered as an additional format, or even as a replacement for Xz.

https://en.wikipedia.org/wiki/Zstandard

**Comment by 89z** [comment]
https://github.com/rust-lang/rustup/issues/2683

**Comment by the8472** [comment]
The last time this was evaluated (#21724) was that xz provided better compression than even very high zstd settings.

**Comment by 89z** [comment]
That may be true, but Compression isnt and shouldnt be the only factor.

Using Zstandard can result in an order of magnitude decrease in decompression speed. That is not an exaggeration:

https://github.com/rust-lang/rustup/issues/2488#issuecomment-692563689

So if you consider the sum of:

1. Time to download
2. Time to extract

Zstandard wins hands down. The downside will be yes, 8% - 10% larger archives, but I think most people would agree its well worth the tradeoff.

**Comment by the8472** [comment]
People live with constrained internet speeds or even data caps. To some bytes are probably valuable than CPU cycles.

**Comment by 89z** [comment]
Its not about CPU cycles, its about Time.

I apologize that I didnt make that more clear. If you consider the sum of:

1. Time to download
2. Time to extract

Zstandard takes significantly less time.

**Comment by the8472** [comment]
That does not affect my argument. The bytes transferred will also be more valuable than wall-time.

Also, wall-time could probably be optimized by decompressing multiple items in parallel.

**Comment by 89z** [comment]
I think I am doing a poor job of explaining. When a person installs an archive,
they incur two costs:

1. Cost in bytes of archive size
2. Cost in time to download and extract

This is not a point of argument. What is a point of argument, is the respective
value of both. For example, given these options:

option | size   | time
-------|--------|-----
1      | 100 MB | 1 hour
2      | 110 MB | 1 second

I think the average person go with option 2. Now, thats not the situation we have
here, but we certainly do have a large time difference to consider. Here are the
actual numbers:

option | size   | time
-------|--------|--------
Xz     | 122 MB | 9.218 s
Zst    | 99 MB  | 0.839 s

https://github.com/rust-lang/rustup/issues/2488#issuecomment-692563689

I didnt notice it before, but as you can see, Zst is actually faster and smaller.
But even assuming it was 10% bigger, say `134.2 MB`. Are you really saying that
10% offsets the 90% reduction in extract time? If you are saying that, then where
is your limit? Where is the threshold for you where the time savings makes it
worth it?

**Comment by the8472** [comment]
> I didnt notice it before, but as you can see, Zst is actually faster and smaller

Was that with the xz defaults or `xz -9`? The issue I linked previously distinguished those cases and found the latter to be smaller.

> Are you really saying that 10% size offsets the 90% reduction in extract time? If you are saying that, then where is your limit? Where is the threshold for you where the time savings makes it worth it?

That depends on several factors, e.g. how often one performs these actions.

From a time perspective alone (i.e. ignoring the monetary costs of data caps/metered connections) you have to consider the pipelining/stream-processing case. If you are downloading and decompressing at the same time then those 9 seconds can be *entirely* masked by decompressing while downloading.

122 MB / 9.218 s = 13.2MB/s. So anyone with internet *slower* than 105Mbit/s will be unaffected by decompression times.  Improving compression ratio will shorten the wall-time they'll spend as it will decrease the time they will spend downloading.

If these actions are currently performed sequentially then there is massive untapped optimization potential that does not require changing the compression algorithm.

> this is a poor argument, as any tactic like this could also be applied to Zstandard archives.

The gains of that would be far smaller since it already is fast. thus there is an asymmetry between the cases that favors optimizing xz.

**Comment by the8472** [comment]
I notice that rustup has support for multiple formats.

Adding zstd as *additional* format should be fine. But *replacing* xz wouldn't be a good idea be as long as it still provides superior compression which helps bandwidth-constrained users.

**Comment by 89z** [comment]
> Was that with the xz defaults or `xz -9`? The issue I linked previously distinguished those cases and found the latter to be smaller.

This file:

https://static.rust-lang.org/dist/2021-02-11/rustc-1.50.0-x86_64-pc-windows-gnu.tar.xz

is 85.9 MB. If I extract the tar and recompress, here are some results with Zstandard:

level | size
------|--------
22    | 74.2 MB
21    | 86.7 MB

So again, in some cases Zstandard is actually smaller than Xz, while still enjoying massive speed benefits.

> From a time perspective alone (i.e. ignoring the monetary costs of data caps/metered connections) you have to consider the pipelining/stream-processing case. If you are downloading and decompressing at the same time then those 9 seconds can be _entirely_ masked by decompressing while downloading.

Ok, but is RustUp, or whatever tools actually doing this? If not, its probably much easier to just switch to Zstandand than mess with that. Especially considering that by your own admission, RustUp already supports Zstandard, so the work in that regard is already done.

Also again, this ignores the fact that Zstandard would also benefit from any streaming changes.

> Adding zstd as _additional_ format should be fine. But _replacing_ xz wouldn't be a good idea as long as it still provides superior compression which helps bandwidth-constrained users.

I mean personally, it seems wasteful to keep Xz around if Zstandard is added, as I have repeatedly shown that Zstandard is actually smaller in some cases. But in the end I dont really care what happens to Xz as long as Zstandard is added. Part of my motivation for this, is that tooling for Xz is starting to show its age.  For example, the most popular Xz package for the Go language, the owner doesnt want to do updates because they have said essentially, "whats the point with better formats like Zstandard around". So I looked into myself, and I pretty much agree with them.

**Comment by the8472** [comment]
> So again, in some cases Zstandard is actually smaller than Xz, while still enjoying massive speed benefits.

If that is the case then maybe zstd has improved since the last measurements and they need to be redone.

> Also again, this ignores the fact that Zstandard would also benefit from any streaming changes.

That is not really a "benefit", it is an argument that decompression time is not relevant up to the point where it exceeds download time. I.e. for slow connections compression ratio is the sole determining factor for wall-time.

> Ok, but is RustUp, or whatever tools actually doing this? If not, its probably much easier to just switch to Zstandand than mess with that.

If they aren't that needs to be fixed and may actually be simpler than adding logic that determines when to use which format. And it shouldn't be all that complex, even in a shell it can be accomplished with some pipes.

**Comment by 89z** [comment]
You seem content to just hand wave away the massive speed benefits the Zstandard format offers. I find your position strange, its almost like you would be content if no new format were ever made again, and Xz was used for decades into the future. Zstandard was created for a reason, people saw the decompress speed for Xz was lacking.  Because you are not taking these differences seriously, I do not wish to exchange any more comments with you. If anyone else decides to comment, I can pick up the conversation at that time.

This format offers real measurable benefits, not hypothetical ones. You should consider, that perhaps people will use their own tools with these archives, and not official Rust tools, so they may not benefit from these mitigations you are suggesting. It just seems quite strange to me that you would fight so hard against this. I wonder what you motivation is? Is is simply that a change cost effort to do? As was discussed, some of the work has already been done, as the RustUp tool already has Zstandard support.

At any rate, feel free to comment further, but I wont be responding to any further comment from you.

**Comment by the8472** [comment]
> You seem content to just hand wave away the massive speed benefits the Zstandard format offers. 

I do not. I argue that the benefit can be zero or negative for a significant amount of the user-base, conditional on rustup being properly pipelined and xz still offering superior compression speeds.

For some users zstd could be a net-positive. I never denied that.

> its almost like you would be content if no new format were ever made again, and Xz was used for decades into the future.

I never said any of the words you're putting in my mouth here. I guess that's the "almost" in that sentence does a lot of work here, you're trying to read something between the lines, but I assure you it is not there.

> It just seems quite strange to me that you would fight so hard against this

As I clarified in https://github.com/rust-lang/rust/issues/82595#issuecomment-787122802 I am not against adding zstd. I am against removing xz, conditional on the assumption that it still provides superior compression ratio, which may have to be measued again, which I conceded in https://github.com/rust-lang/rust/issues/82595#issuecomment-787129498

> so they may not benefit from these mitigations

They're not mitigations. They're obvious optimizations *if one cares about wall-time*, which you seem to do. If they haven't been done then I guess this just hasn't been a priority so far.
Basically it's something one should do independently of which compression algorithm is chosen.

> I wonder what you motivation is? Is is simply that a change cost effort to do? 

I prefer advancing the pareto frontier rather than merely picking to a different point on a fixed curve.

If zstd now always provides superior compression (TBD) then that's an easy win and it can become the default, no objection.
If it doesn't then the situation gets a lot more complicated and it only makes sense in some situations and not in others. Therefore either one should pluck lower-hanging optimization fruit or a plan when to pick which option will be necessary.


> At any rate, feel free to comment further, but I wont be responding to any further comment from you.

It's rather unfriendly to ragequit a discussion while throwing insinuations and inferences into the room.

**Comment by kinnison** [comment]
For reference, when `rustup` was discussing adding `.zst` support, @joshtriplett did some very basic tests on compression size which he detailed on the issue in `rustup` - At least one data point showed `.zst` being significantly smaller than `.xz` so there may be merit in considering `.zst` from a size/bandwidth perspective as well as a decompression speed perspective.  That comment is here: https://github.com/rust-lang/rustup/issues/2488#issuecomment-692563689

**Comment by 89z** [comment]
https://github.com/rust-lang/rustup/issues/1858

**Comment by 89z** [comment]
closing for lack of interest
